framework/src/ble_manager/ble_manager_message.c

```diff
@@ -0,0 +1,180 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <stdio.h>
+#include <debug.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+#include <tinyara/lwnl/lwnl.h>
+#include "ble_manager_log.h"
+#include "ble_manager_event.h"
+#include "ble_manager_message.h"
+#include "ble_manager_msghandler.h"
+
+#define BLEMGR_MSG_QUEUE_NAME "/dev/blemgr_msg"
+#define BLEMGR_MSG_TAG_ERR "[LE MSG] "
+#define BLE_MESSAGE_ERROR                                                        \
+	do {                                                                     \
+		BLE_LOG_ERROR(BLEMGR_MSG_TAG_ERR"%s\t%s:%d\n", __FUNCTION__, __FILE__, __LINE__); \
+		BLE_LOG_ERROR(BLEMGR_MSG_TAG_ERR"err num(%d)\n", errno);                                \
+	} while (0)
+
+static inline int _send_message(int fd, void *buf, int buflen)
+{
+	int sent = 0;
+	while (1) {
+		int res = write(fd, (void *)buf + sent, buflen - sent);
+		if (res < 0) {
+			BLE_MESSAGE_ERROR;
+			return -1;
+		}
+		sent += res;
+		if (sent == buflen) {
+			break;
+		}
+	}
+	return 0;
+}
+
+static inline int _recv_message(int fd, void *buf, int buflen)
+{
+	int received = 0;
+	while (1) {
+		int res = read(fd, buf + received, buflen - received);
+		if (res < 0) {
+			BLE_MESSAGE_ERROR;
+			return -1;
+		}
+		received += res;
+		if (received == buflen) {
+			break;
+		}
+	}
+	return 0;
+}
+
+/*
+ * Public functions
+ */
+int blemgr_message_in(ble_handler_msg *msg, ble_handler_queue *queue)
+{
+	int fd = open(BLEMGR_MSG_QUEUE_NAME, O_WRONLY);
+	if (fd < 0) {
+		BLE_MESSAGE_ERROR;
+		return -1;
+	}
+
+	int res = _send_message(fd, (void *)msg, sizeof(ble_handler_msg));
+	if (res < 0) {
+		BLE_MESSAGE_ERROR;
+		close(fd);
+		return -1;
+	}
+
+	close(fd);
+
+	return 0;
+}
+
+
+// it's not thread-safe. in other words only one thread should call
+// this function
+int blemgr_message_out(ble_handler_msg *msg, ble_handler_queue *queue)
+{
+	fd_set rfds = queue->rfds;
+	int res = select(queue->max + 1, &rfds, NULL, NULL, NULL);
+	if (res <= 0) {
+		if (errno == EINTR) {
+			return 1;
+		}
+		BLE_MESSAGE_ERROR;
+		return -1;
+	}
+	if (FD_ISSET(queue->fd, &rfds)) {
+		res = _recv_message(queue->fd, (void *)msg, sizeof(ble_handler_msg));
+		if (res < 0) {
+			BLE_MESSAGE_ERROR;
+		}
+		blemgr_msg_s *bmsg = msg->msg;
+		bmsg->result = blemgr_handle_request(bmsg);
+		if (msg->signal) {
+			sem_post(msg->signal);
+		}
+	}
+#ifdef CONFIG_LWNL80211
+	if (FD_ISSET(queue->nd, &rfds)) {
+		res = lwnl_fetch_ble_event(queue->nd, (void *)msg, sizeof(ble_handler_msg));
+		if (res < 0) {
+			BLE_MESSAGE_ERROR;
+		}
+		blemgr_msg_s *bmsg = msg->msg;
+		bmsg->result = blemgr_handle_request(bmsg);
```

Comment:
I think we'd better check `msg` and `bmsg` are NULL or not.
Because they can be dereferencing.

Reply:
'msg' cannot be NULL forever.
It defined just before calling "blemgr_message_out".